### PR TITLE
feat: support multiple npm registries and tokens via .npmrc

### PR DESCRIPTION
**Pull request by pedrobarco**
This PR adds the ability to support multiple registries and auth tokens via .npmrc.

This branch was cloned from #499 and addresses #502. 

It is being developed in an upstream branch because the e2e test won't run on my fork since it requires a secret which isn't available for forks.

### CODE REVIEW

File: npm/private/npm_translate_lock.bzl [hunk 1]
```diff
@@ -277,36 +330,14 @@ def _gen_npm_imports(lockfile, root_package, attr):
...
-re-generating the lockfile should generate a fully qualified tarball URL for this package.
-
-To disable this warning, set `warn_on_unqualified_tarball_url` to False in your
-`npm_translate_lock` repository rule.
-====================================================================================================
-
-""".format(name, version, tarball))
+                    (scope, _) = utils.parse_package_name(name)
+                    registry = "https://{}".format(registries[scope]) if scope in registries else utils.npm_registry_url
                 url = registry + tarball
```
Comment (gregmagolan) [inline npm/private/npm_translate_lock.bzl:341]: should make sure that `registry` ends with `/` here when concatenating the strings `url = registry + tarball` or else inject `/` separator if between them if it doesn't

File: npm/private/utils.bzl [hunk 2]
```diff
@@ -117,6 +117,26 @@ def _is_at_least_bazel_6():
     # native.bazel_version only works in repository rules.
     return "apple_binary" not in dir(native)
 
+def _parse_package_name(package):
+    # Parse a @scope/name string and return a [scope, name] list
```
Comment (gregmagolan) [inline npm/private/utils.bzl:121]: nit: "return a (scope, name) tuple"

File: npm/private/npm_translate_lock.bzl [hunk 3]
```diff
@@ -132,35 +133,87 @@ def _gather_values_from_matching_names(keyed_lists, *names):
...
-        if _NPM_TOKEN_KEY in npmrc:
-            # Parse environment variable from config
-            token = npmrc[_NPM_TOKEN_KEY]
+    Args:
+        npmrc: The `.npmrc` file.
+        npmrc_path: The file path to `.npmrc`.
+        environ: A map of environment variables with their values.
+
+    Returns:
+        A list with a tokens dict and a registries dict.
```
Comment (gregmagolan) [inline npm/private/npm_translate_lock.bzl:174]: nit: A tuple with a tokens dict and a registries dict

**Comment by gregmagolan** [comment]
Looks great. Just a few minor comments and this is good to land. Will cut a minor release after it does.

File: npm/private/utils.bzl (hunk 2 above)
Comment (pedrobarco) [reply npm/private/utils.bzl:121]: My bad, I've had it like you suggested but I found a similar function that was using the same description: https://github.com/aspect-build/rules_js/blob/724c377bfed9db82881b4174b9652700f9c426c7/npm/private/utils.bzl#L33-L39

Should I change its description too?

File: npm/private/npm_translate_lock.bzl (hunk 3 above)
Comment (pedrobarco) [reply npm/private/npm_translate_lock.bzl:174]: Same as above. On it ⛵

File: npm/private/utils.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/utils.bzl:121]: Yes please. Nice catch. Looks like that one got out of sync.

**Review by gregmagolan — APPROVED** [review]
🌮 🚀

**Pull request merged**